In 2024 CQC required registration for API key.

simondashnash opened the pull request:
Hi Evan

I had a go at updating your library to work with CQC's API key. Please could you review and merge if its OK?

This commit adds a file subscription_key.R which replaces partner_code.R and amends some other files to use a new base URL and pass the API key in the header. I also amended the introduction.Rmd file to reflect CQC's change.

Yours

Simon

simondashnash committed:
In 2024 CQC required registration for API key.
This commit adds a file subscription_key.R which replaces partner_code.R and amends multiple other files to use a new base URL and pass the API key in the header.

evanodell commented on R/utils-query-construct.R:
```diff
@@ -28,3 +28,32 @@ cqc_query_construction <- function(query) {
+
+#   suppressMessages(cont <- jsonlite::fromJSON(httr::content(x, "text")))
+
+#   cont
```
Instead of commenting out old formulas can you remove them entirely?

evanodell reviewed (commented):
Instead of commenting out old functions you can remove them entirely. 

Otherwise it looks good to me. I've not been actively maintaining any of my R packages so I'm rusty on the code but it seems to work. 

Also please add yourself as a contributor in the DESCRIPTION file and bump the version up.

simondashnash committed:
removed commented out old function

simondashnash committed:
removed commented out old base url

simondashnash committed:
Tidied up amendments to work with CQC's new API key requirement

simondashnash commented:
Hi.
I tidied up commented lines, added contributor lines to DESCRIPTION and ran devtools::document() . Hopefully OK now.

Simon

simondashnash committed:
Better handling of errors on get details.

simondashnash committed:
Many tests passing. Now need to add api key to the reports.R which has its own httr

simondashnash committed:
reports tests now pass

simondashnash committed:
invalid location returns a df with error message, test now looks for this

simondashnash committed:
minor changes to introduction file

simondashnash committed:
incremented version number

simondashnash commented:
I amended the tests to prompt user for API key, did some other tidying as you advised, added myself as a contributor and incremented the version number.

simondashnash committed:
Retry transient API failures and raise structured errors instead of silent malformed data

cqc_query_construction() previously swallowed non-200 responses by
building an unused local error data.frame and implicitly returning it,
with no logging and no retry. Downstream code (e.g. cqc_get_data's
`cont$totalPages` check) then crashed with a cryptic
"argument is of length zero" instead of a useful error, and per-location
lookups (cqc_location_details) silently dropped failed items with no
record.

Now retries up to 3 times with exponential backoff (the CQC API has
frequent transient blips), and raises a structured cqcr_api_error
condition (status/url/body) if all retries fail, so callers can
tryCatch() on it and log/skip individual failures deliberately.
